src/rpdk/guard_rail/rule_library/core/schema-linter-core-rules.guard

```diff
@@ -27,67 +20,132 @@ rule ensure_primary_identifier_exists_and_not_empty
+                <<
+                {
+                    "result": "NON_COMPLIANT",
+                    "check_id": "PID003",
+                    "message": "primaryIdentifier MUST be either readOnly or createOnly"
+                }
+                >>
+            }
+        }
+
+        when createOnlyProperties !exists {
+            primaryIdentifier[*] {
+                this IN %read
+                <<
+                {
+                    "result": "NON_COMPLIANT",
+                    "check_id": "PID003",
+                    "message": "primaryIdentifier MUST be either readOnly or createOnly"
+                }
+                >>
+            }
+        }
+    }
+
+    when readOnlyProperties !exists {
+        createOnlyProperties !exists
         <<
         {
             "result": "NON_COMPLIANT",
-            "check_id": "PID003",
+            "check_id": "PID00",
             "message": "primaryIdentifier MUST be either readOnly or createOnly"
         }
         >>
+
+        when createOnlyProperties exists {
+            let create = createOnlyProperties
+            primaryIdentifier[*] {
+                this IN %create
+                <<
+                {
+                    "result": "NON_COMPLIANT",
+                    "check_id": "PID003",
+                    "message": "primaryIdentifier MUST be either readOnly or createOnly"
+                }
+                >>
+            }
+        }
     }
 }
 
 
 
 rule ensure_create_and_read_only_intersection_is_empty
 {
-    readOnlyProperties[*] {
-        this !IN %create
-        <<
-        {
-            "result": "NON_COMPLIANT",
-            "check_id": "PR001",
-            "message": "read/createOnlyProperties MUST NOT have common properties"
+    when readOnlyProperties exists {
+        when createOnlyProperties exists {
+            let create = createOnlyProperties
+            readOnlyProperties[*] {
+                this !IN %create
+                <<
+                {
+                    "result": "NON_COMPLIANT",
+                    "check_id": "PR001",
+                    "message": "read/createOnlyProperties MUST NOT have common properties"
+                }
+                >>
+            }
         }
-        >>
     }
 
-    createOnlyProperties[*] {
-        this !IN %read
-        <<
-        {
-            "result": "NON_COMPLIANT",
-            "check_id": "PR002",
-            "message": "create/readOnlyProperties MUST NOT have common properties"
+    when createOnlyProperties exists {
+        when readOnlyProperties exists {
+            let read = readOnlyProperties
+            createOnlyProperties[*] {
+                this !IN %read
+                <<
+                {
+                    "result": "NON_COMPLIANT",
+                    "check_id": "PR001",
+                    "message": "read/createOnlyProperties MUST NOT have common properties"
+                }
+                >>
+            }
```

Comment:
Is there a reason we declare scenario with the conditions reversed?

Logically, wouldn't it be sufficient to just check either 1 of them to detect this error?

Reply:
yea, I think just one is enough